Game of Life Memory Leak Fix + Other Changes

Brandon502 opened the pull request:
Fixed memory leak using new. Removed gameoflife struct and reverted back to using pointers to segenv.data (similar to starLeds version). 

Added Blur slider. Blurs dead cells slowly to bgColor depending on slider instead of immediately dying. Doesn't effect game logic.

Removed confusing pauseFrames, uses .step to pause now.

Keeps track of last palette used and immediately recolors cells if new palette/color selected. Can be improved.

Changes gameSpeed slider to be more consistent. Speed = updates/second now in increments of 4.

Brandon502 committed:
GoL Memory Leak fix

Removed gameoflife struct.

Brandon502 committed:
GoL - Added blur slider

Blurs dead cells instead of immediately turning off depending on slider value. Blur is disabled when using overlay.

Brandon502 committed:
GoL - Removed pauseFrames

Uses .step now to pause.

Brandon502 committed:
GoL - Change default values.

General cleanup.

Brandon502 committed:
GoL - Palette Switching

Recolors all live cells on palette/color change. Before you needed a game restart or mutation over time.

Brandon502 committed:
GoL - Changed speed slider to updates/sec

Every 4 ticks of speed adds 1 update per second. 1 - 64.

Brandon502 committed:
Merge branch 'MoonModules:mdev' into mdev

Brandon502 committed:
Merge branch 'mdev' of https://github.com/Brandon502/WLED into mdev

Brandon502 committed:
GoL - Uses cIndex more consistently

Brandon502 committed:
Merge remote-tracking branch 'upstream/mdev' into mdev

softhack007 commented:
Hi, thanks for this update.
I'll look into the details like game speed in the next days, and then we can merge it into the code for the new release of WledMM (it's still the same "upcoming" release) 

> I couldn't figure out a way to store the actual previous palette used, so it cheats by just using the first color of the palette

You could use `uint8_t currentPal = SEGMENT.palette;` to get the current palette ID.
Downside: sometimes a palette is dynamically adjusted, while the ID stays the same. For example: during effect blending, or when using one of the audioreactive palettes. So this would also not make the code 100% bullet proof.

But your "cheat" is OK for me, too. I doesn't need to be perfect.

Brandon502 commented:
`uint8_t currentPal = SEGMENT.palette;`

Some how overlooked this. Didn't think about palettes that change overtime, keeping the first color would work better for those. Each method seems to have a slight drawback, implementing both would be overkill. I'll have to test out those special palettes. 

I have a few more changes I've made in StarLeds that I need to convert to WLED and push. Finally figured out a way to make blending look better when a game starts repeating.

Brandon502 committed:
Merge branch 'MoonModules:mdev' into mdev

Brandon502 committed:
GoL - Rework

Blur now fades while game is paused.
New blur mode when blur slider is >220
Added variables to make code easier to follow instead of check1 custom1 etc. (not needed)
Removed % with random8 calls.
Misc changes

Brandon502 committed:
Merge branch 'mdev' of https://github.com/Brandon502/WLED into mdev

Brandon502 commented:
I'll post videos when I get a chance later. Code is a bit more complicated, especially the new redraw loop. Added new blend option when blur slider >220. There is an unintended effect when you have two segments and overlay off. Looks really good on high game speeds, but breaks (flashes) on lower ones. Tried a few ways to get it working on lower speeds but no luck. 

Default parameters don't seem to be working. May just be using progmem incorrectly. 

futureCells is technically not needed to be stored in memory. It is only needed when game updates occur.

Brandon502 commented:
Short video showing the new options. 

- Top left: Blur >220. The background keeps a faded color of the last alive cell on that led. Kind of a history/turf war of the alive colors. 
- Top right: High blur value with multiple palette changes to show instant switching and blur fading as the game ends. 
- Bottom left: Unintended effect when you turn off overlay when you have another effect in the background. Alive cells take the background color and you can apply a different blur color. Video shows white then switching to Atlantica. 
- Bottom right: Side effect when game speed is too low, tons of flashing since there are effects stacked. May be able to figure out how to prevent it so there would be a second overlay type without a toggle.


https://github.com/MoonModules/WLED/assets/105077712/b96e22b7-d23b-4d92-b382-e5b096a5524f

Brandon502 committed:
GoL - Allows default overlay mode.

Flicker fixed by always redrawing dead cells if overlay unchecked. Can swap between overlay types by checking/unchecking overlayBG.

Brandon502 committed:
GoL - gliderLength switched to uint16_t

Brandon502 committed:
GoL - Correct data allocation

Fixed overlooked data allocation after switching gliderLength to uint16.

Brandon502 commented:
Stress tested on a 64x32 matrix with 4 pins. Showing 30 FPS with max game speed. 40+ when updating is set to 20/sec.


https://github.com/MoonModules/WLED/assets/105077712/509842ee-bd31-4ba5-af87-ae236eb5d219

Brandon502 committed:
GoL - Small color fix and glider detection change.

Wrap is disabled if only 5 remaining alive cells (Glider).

Brandon502 committed:
Merge remote-tracking branch 'upstream/mdev' into mdev

softhack007 commented on wled00/FX.cpp:
```diff
@@ -5233,27 +5230,34 @@ uint16_t mode_2Dgameoflife(void) { // Written by Ewoud Wijma, inspired by https:
+    cIndex = y * cols + x;
+    bool alive = getBitValue(cells, cIndex);
+    if (alive) aliveCount++;
+    CRGB color = SEGMENT.getPixelColorXY(x,y);
```
style: this variable "shadows" the global color defined on line 5194. Maybe better to use a "unique" name.

softhack007 commented on wled00/FX.cpp:
```diff
@@ -5164,66 +5164,63 @@ static void setBitValue(uint8_t* byteArray, size_t n, bool value) {
-        SEGMENT.setPixelColorXY(x,y,!SEGMENT.check1?color : RGBW32(color.r, color.g, color.b, 0));
+        SEGMENT.setPixelColorXY(x,y, allColors ? random16() * random16() : color);
       }
+      else if (!overlayBG) SEGMENT.setPixelColorXY(x,y, allColors ? RGBW32(bgColor.r, bgColor.g, bgColor.b, 0) : bgColor); // set background color if not overlaying
```
maybe i missed something - it seems like "then" and "else" sides of the `?` operator are the same? 
because CRGB does not have a "white" channnel.

Brandon502 replied:
I agree they're the same. This was code I kept from the original version since I didn't know if it was required for RGBW setups that I couldn't test on.

softhack007 replied:
Good point about RGBW. I don't have any rgbw panels either, however I'll do some testing by changing my "virtual" 52x52 to sk6812 and checking preview.

softhack007 commented on wled00/FX.cpp:
```diff
@@ -5233,27 +5230,34 @@ uint16_t mode_2Dgameoflife(void) { // Written by Ewoud Wijma, inspired by https:
+    if (palChanged && alive) SEGMENT.setPixelColorXY(x,y, SEGMENT.color_from_palette(random8(), false, PALETTE_SOLID_WRAP, 0)); // Recolor alive cells
+    else if (overlayBG & alive) SEGMENT.setPixelColorXY(x,y, allColors ? RGBW32(color.r, color.g, color.b, 0) : color);         // Redraw alive cells for overlayBG
+    if (palChanged && !alive && !overlayBG) SEGMENT.setPixelColorXY(x,y, bgColor);                                              // Remove blurred cells from previous palette
+    else if (blurDead && !alive) SEGMENT.setPixelColorXY(x,y, blend(color, bgColor, blur));                                     // Blur dead cells (paused)
```
optimization: you could use `color_blend()` instead of `blend()`

https://github.com/MoonModules/WLED/blob/fc173b3bc00694e59b653ca230133052b5476c05/wled00/fcn_declare.h#L53

The WLED function color_blend() is slightly faster because it can handle all color components in "one shot".
You only have to use `uint32_t` instead of `CRGB` for colors. `uint32_t` (aka RGBW32) is the "WLED native" format for colors. 

In case you need to convert from CRGB to RGBW32, its necessary to mask off the highest byte after conversion: `uint32_t wled_color = uint32_t(fastled_color) & 0x00FFFFFF`. Because fastled understands 32bit as "RGBA" (a=alpha) and not "RGBW" (w=white).

Brandon502 replied:
I'll look into using uint32_t instead of CRGB along with color_blend(). Thanks for reviewing everything.

softhack007 commented on wled00/FX.cpp:
```diff
@@ -5263,78 +5267,81 @@ uint16_t mode_2Dgameoflife(void) { // Written by Ewoud Wijma, inspired by https:
-      if (!SEGMENT.check2) SEGMENT.setPixelColorXY(x,y, !SEGMENT.check1?backgroundColor : RGBW32(backgroundColor.r, backgroundColor.g, backgroundColor.b, 0));
+      setBitValue(futureCells, cIndex, false);
+      // Blur/turn off dying cells
+      if (!overlayBG) SEGMENT.setPixelColorXY(x,y, blend(SEGMENT.getPixelColorXY(x,y), bgColor, bgBlendMode ? bgBlur : blur));
```
optimization: try to use color_blend() instead of blend()

softhack007 commented on wled00/FX.cpp:
```diff
@@ -5233,27 +5230,34 @@ uint16_t mode_2Dgameoflife(void) { // Written by Ewoud Wijma, inspired by https:
+    if (alive) aliveCount++;
+    CRGB color = SEGMENT.getPixelColorXY(x,y);
+    if (palChanged && alive) SEGMENT.setPixelColorXY(x,y, SEGMENT.color_from_palette(random8(), false, PALETTE_SOLID_WRAP, 0)); // Recolor alive cells
+    else if (overlayBG & alive) SEGMENT.setPixelColorXY(x,y, allColors ? RGBW32(color.r, color.g, color.b, 0) : color);         // Redraw alive cells for overlayBG
```
same question - is there a difference between `RGBW32(color.r, color.g, color.b, 0)` and `color`?

softhack007 replied:
minor (style): 'overlayBG' is used in a bitwise '&' operation. Might be better to use '&&'

softhack007 reviewed (changes requested):
Hi, thanks for the update.
i've tested your new code quickly, and it works for me. 

Below some recommendations that could slightly improve speed.

Brandon502 commented:
Is this a known issue regarding overlays "breaking"?

Normal overlay:

https://github.com/MoonModules/WLED/assets/105077712/b2abfa95-27de-4382-9f70-6a8bed544e0d

Glitched overlay:

https://github.com/MoonModules/WLED/assets/105077712/89591222-101b-42e8-b9f6-2e44133c975d

Seems like the segments start sharing colors instead of keeping them separate. It allows for some neat effects. Just makes debugging annoying when I don't want it to happen.


https://github.com/MoonModules/WLED/assets/105077712/b258c860-a12d-4192-85b5-1b9a2720d335

Brandon502 committed:
GoL CRGB changed to uint32 & color_blend()

Other small changes.

softhack007 commented on wled00/FX.cpp:
```diff
@@ -5263,78 +5267,81 @@ uint16_t mode_2Dgameoflife(void) { // Written by Ewoud Wijma, inspired by https:
-    } 
+    }
+    else { // blur dead cells 
+      if (!cellValue && !overlayBG && !bgBlendMode) SEGMENT.setPixelColorXY(x,y, blend(SEGMENT.getPixelColorXY(x,y), bgColor, bgBlendMode ? bgBlur : blur)); 
```
minor: this line can be simplified a bit, because `bgBlendMode` must be false due to `if (!cellValue && !overlayBG && !bgBlendMode)`

softhack007 commented:
> Is this a known issue regarding overlays "breaking"?

Normally not - only two exceptions that I know about
* "Use global LED buffer" is enabled, making all pixels shared (plus known bug https://github.com/MoonModules/WLED/issues/58)
* effects run with 1D->2D mapping (pinwheel, bar, etc.) In this case the segment buffer cannot be used.

Brandon502 committed:
GoL - Small fix

Remove redundant check

Brandon502 commented:
@softhack007 I think I made all your recommended changes. Although I am not sure when to use `uint32_t wled_color = uint32_t(fastled_color) & 0x00FFFFFF` conversion. Currently just set all CRGB to uint32. Everything seems to work, using color_blend didn't have a noticeable affect on FPS on my end, ~+1 FPS.

Brandon502 committed:
GoL - Game start transition

Starting alive cells are randomly colored in and bg fades on game start.

Brandon502 committed:
GoL - Redraw loop changes

Merged generation == 1 checks into standard checks.

Brandon502 commented:
Finally got alive cells to spawn in randomly on game start working on everything, but overlayBG. Had a small flicker on start so decided to remove it. Looks a lot better with transition between games. The redraw loop is even more complicated now unfortunately.

softhack007 reviewed (approved):
Thanks, looks good 👍 

Just tell once you're happy with the code, and i'll merge it into mdev.

Brandon502 commented:
> Thanks, looks good 👍
> 
> Just tell once you're happy with the code, and i'll merge it into mdev.

Sounds good, I'll play around with it some more for a day or two to make sure everything is working normally and no bugs pop up. I'll let you know soon.

Brandon502 commented:
@softhack007 Haven't run into any bugs on my setup so far. Feel free to merge whenever.
